astropy/utils/masked/tests/test_functions.py

```diff
@@ -517,3 +518,64 @@ def test_lexsort_mix(self, axis):
         assert_array_equal(mbma_lexsort, expected_ba)
         mbma_lexsort2 = np.lexsort(np.stack([mb, self.a], axis=0), axis=axis)
         assert_array_equal(mbma_lexsort2, expected_ba)
+
+
+class TestStructuredUfuncs:
+    """Test with structure dtypes, using erfa ufuncs."""
+
+    def test_erfa_d2tf_tf2d(self):
+        mask = np.array([True, False, False])
+        days = Masked([0.25, 0.875, 0.0625], mask=mask)
+        sign, ihmsf = erfa.d2tf(3, days)
+        assert_array_equal(sign.mask, mask)
+        for name in ihmsf.dtype.names:
+            assert_array_equal(ihmsf[name].mask, mask)
+
+        # Check roundtrip.
+        check = erfa.tf2d(sign, ihmsf["h"], ihmsf["m"], ihmsf["s"] + ihmsf["f"])
+        assert_allclose(check.unmasked, days, atol=1e-3 / 24 / 3600)
+        assert_array_equal(check.mask, mask)
+
+    def test_erfa_astrom(self):
+        mask = np.array([True, False, False])
+        jd2 = Masked([0, 0.401182685, 0.5], mask=mask)
+        astrom, eo = erfa_ufunc.apci13(2456165.5, jd2)
+        assert_array_equal(eo.mask, mask)
+        for n in astrom.dtype.names:
+            # .T for multi-element fields.
+            assert np.all(astrom[n].mask.T == mask)
+
+        along = np.array([0.125, 0.25, 0.35])
+        # Not going to worry about different masks for different elements.
+        # In principle aper could propagate just what it needs.
+        astrom["along"] = Masked([0.125, 0.25, 0.35], mask)
+        astrom2 = erfa_ufunc.aper(Masked(np.ones(3), [False, True, False]), astrom)
+        assert_array_equal(astrom2["eral"].unmasked, along + 1.0)
+        mask2 = mask | [False, True, False]
+        for n in astrom2.dtype.names:
+            # .T for multi-element fields.
+            assert np.all(astrom2[n].mask.T == mask2)
+
+    def test_erfa_atioq(self):
+        # Regression test for gh-16123, using test from erfa.
+        astrom, _ = erfa_ufunc.apio13(
+            2456384.5,
+            0.969254051,
+            0.1550675,
+            -0.527800806,
+            -1.2345856,
+            2738.0,
+            2.47230737e-7,
+            1.82640464e-6,
+            731.0,
+            12.8,
+            0.59,
+            0.55,
+        )
+        astrom = Masked(astrom)
+        ri = 2.710121572969038991
+        di = 0.1729371367218230438
+        aob, zob, hob, dob, rob = erfa_ufunc.atioq(ri, di, astrom)
+        assert isinstance(aob, Masked)
+        assert_allclose(aob, 0.9233952224895122499e-1, atol=1e-12)
+        # Rest presumably is fine too.
```

Comment:
This comment feels like you're asking for trouble 😅

Reply:
OK, let me copy it!